packages/gatsby/src/bootstrap/load-plugins.js

```diff
@@ -10,6 +10,78 @@ const nodeAPIs = require(`../utils/api-node-docs`)
 const testRequireError = require(`../utils/test-require-error`)
 const report = require(`gatsby-cli/lib/reporter`)
 
+// Given a plugin object and a moduleName like `gatsby-node`, check that the
+// path to moduleName can be resolved.
+const resolvePluginModule = (plugin, moduleName) => {
+  let resolved = false
+  try {
+     resolved = require(`${plugin.resolve}/${moduleName}`)
+  } catch (err) {
+    if (!testRequireError(moduleName, err)) {
+      // ignore
+    } else {
+      report.panic(`Error requiring ${plugin.resolve}/${moduleName}.js`, err)
+    }
+  }
+  return resolved
+}
+
+// Given a plugin object, an array of the API names it exports and an
+// array of valid API names, return an array of invalid API exports.
+const getBadExports = (plugin, pluginAPIKeys, apis) => {
+  let badExports = []
+  // Discover any exports from plugins which are not "known"
+  badExports = badExports.concat(
+    _.difference(pluginAPIKeys, apis).map(e => {
+      return {
+        exportName: e,
+        pluginName: plugin.name,
+        pluginVersion: plugin.version,
+      }
+    })
+  )
+  return badExports
+}
+
+const getBadExportsMessage = (badExports, apis) => {
+  const { stripIndent } = require(`common-tags`)
+  const stringSimiliarity = require(`string-similarity`)
+
+  let message = `\n`
+  message += stripIndent`
+    Your plugins must export known APIs from their gatsby-node.js.
+    The following exports aren't APIs. Perhaps you made a typo or
+    your plugin is outdated?
+
+    See https://www.gatsbyjs.org/docs/node-apis/ for the list of Gatsby Node APIs`
+
+  badExports.forEach(bady => {
+    const similarities = stringSimiliarity.findBestMatch(
+      bady.exportName,
+      apis
+    )
+    message += `\n — `
+    if (bady.pluginName == `default-site-plugin`) {
+      message += `Your site's gatsby-node.js is exporting a variable named "${
```

Comment:
This message should vary now based on whether its node/ssr/browser apis right?

Reply:
Oh yes, it should 👍